src/gap_to_julia.jl

```diff
@@ -22,233 +22,205 @@ of converted objects and should never be given by the user.
 ## (::Type{Array{T,1}}, :: GapObj) is invoked, with T = Tuple{Int64}; this then
 ## invokes gap_to_julia recursively with signature (::Tuple{Int64},::Any),
 ## which ends up selecting the method below.
-function gap_to_julia(t::T, x::Any) where T <: Type
-    if ! (typeof(x) <: t)
-        throw(MethodError("Wrong type: expected " * string(t) * ","))
-    end
-    return x
-end
-
-## Default
-gap_to_julia(::Type{Any}, x::GapObj) = gap_to_julia(x)
-gap_to_julia(::Type{Any}, x::Any) = x
-gap_to_julia(::T,         x::Nothing) where T <: Type = nothing
-gap_to_julia(::Type{Any}, x::Nothing) = nothing
-
-## Integers
-gap_to_julia(::Type{Int128} ,x::Int64) = trunc(Int128 ,x)
-gap_to_julia(::Type{Int64}  ,x::Int64) = x
-gap_to_julia(::Type{Int32}  ,x::Int64) = trunc(Int32  ,x)
-gap_to_julia(::Type{Int16}  ,x::Int64) = trunc(Int16  ,x)
-gap_to_julia(::Type{Int8}   ,x::Int64) = trunc(Int8   ,x)
-
-## Unsigned Integers
-gap_to_julia(::Type{UInt128},x::Int64) = trunc(UInt128,x)
-gap_to_julia(::Type{UInt64} ,x::Int64) = trunc(UInt64 ,x)
-gap_to_julia(::Type{UInt32} ,x::Int64) = trunc(UInt32 ,x)
-gap_to_julia(::Type{UInt16} ,x::Int64) = trunc(UInt16 ,x)
-gap_to_julia(::Type{UInt8}  ,x::Int64) = trunc(UInt8  ,x)
-
-## BigInts
-gap_to_julia(::Type{BigInt}, x::Int64) = BigInt( x )
-
-function gap_to_julia(::Type{BigInt}, x::GapObj)
-    ## Check for correct type
-    if ! Globals.IsInt(x)
-        throw(ArgumentError("GAP object is not a large integer"))
-    end
+
+struct ConversionError <: Base.Exception
+    obj
+    jl_type
+end
+
+Base.showerror(io::IO, e::ConversionError) =
+    print(io, "failed to convert GapObj to $(e.jl_type):\n $(e.obj)")
+
+function Base.Int64(obj::GapObj)
+    Globals.IsInt(obj) && return Int64(BigInt(obj))
+    throw(ConversionError(obj, Int))
+end
+
+function Base.Float64(obj::GapObj)
+    Globals.IsFloat(obj) && return ValueMacFloat(obj)::Float64
```

Comment:
There are many different float implementations in GAP. `ValueMacFloat` is only defined for `MacFloats`, i..e., machine floats.
```suggestion
    Globals.IsIEEE754FloatRep(obj) && return ValueMacFloat(obj)::Float64
```